src/main/java/org/truffleruby/core/time/RubyDateFormatter.java

```diff
@@ -655,18 +655,21 @@ public static Rope formatToRopeBuilderFast(Token[] compiledPattern, ZonedDateTim
 
                 case FORMAT_NANOSEC: {
                     final int nano = dt.getNano();
-                    final LazyIntRope nanoRope = new LazyIntRope(nano);
+                    final LazyIntRope microSecondRope = new LazyIntRope(nano / 1000);
 
                     final int length = 6;
```

Comment:
It’s a bit surprising to see all this 6-digit / microsecond business happening inside the `FORMAT_NANOSEC` case, which is presumably how this bug snuck through in the first place. Is there anything we can do here (e.g. a commit message, comment, or local naming) to make it clearer what’s going on?

I think the explanation is: although [the default for `%N` is 9 digits](https://github.com/ruby/ruby/blob/v3_0_1/strftime.c#L778-L787) (`N` = “nanoseconds”), our `Time#strftime` fast path is [taken when using 6 digits](https://github.com/oracle/truffleruby/blob/2d75a9837e3f6ca3ffc638bc312212e195e9d7f7/src/main/java/org/truffleruby/core/time/RubyDateFormatter.java#L588-L590) because it’s designed to [improve the performance](https://github.com/oracle/truffleruby/pull/2361) of `Logger#add` specifically, whose [default format](https://github.com/ruby/ruby/blob/v3_0_1/lib/logger/formatter.rb#L21-L23) is `"%Y-%m-%dT%H:%M:%S.%6N "`.

Reply:
Prior to seeing this GitHub comment, I had added a code comment indicating the whole case only handles the `%6N` format string. I think there's some value in maintaining structural parity with the generic case, so I wouldn't want to start creating new token types (e.g., `FORMAT_MICROSEC`), even if that might be clearer.

Please let me know if you think it's clear enough. The bit you have about the context for the fast-path call is helpful, but I'd probably expect to see that info provided for the entire method call.

Reply:
I agree that structural parity is the more important concern. The comments are clear, thank you! The big picture context (i.e. this is all in service of `Logger#add`) remains implicit but there’s now enough information here for someone to more easily discover it.